jmh-samples/src/main/java/org/openjdk/jmh/samples/JMHSample_39_MemoryAccess.java

```diff
@@ -0,0 +1,121 @@
+ * modification, are permitted provided that the following conditions are met:
+ *
+ *  * Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions and the following disclaimer.
+ *
+ *  * Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ *  * Neither the name of Oracle nor the names of its contributors may be used
+ *    to endorse or promote products derived from this software without
+ *    specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.openjdk.jmh.samples;
+
+import org.openjdk.jmh.annotations.*;
+import org.openjdk.jmh.runner.Runner;
+import org.openjdk.jmh.runner.RunnerException;
+import org.openjdk.jmh.runner.options.Options;
+import org.openjdk.jmh.runner.options.OptionsBuilder;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Random;
+import java.util.concurrent.TimeUnit;
+
+@BenchmarkMode(Mode.AverageTime)
+@OutputTimeUnit(TimeUnit.MILLISECONDS)
+@Warmup(iterations = 5, time = 1)
+@Measurement(iterations = 5, time = 1)
+@Fork(5)
+@State(Scope.Benchmark)
+public class JMHSample_39_MemoryAccess {
+    public static final int N = 200_000_000;
+
+    /*
+     * This example highlights the pitfall of accidentally measuring memory access instead of processing time.
+     *
+     * An int array has got a different memory layout than an ArrayList of boxed ints.
+     * This can lead to useless results because the memory access is completely different.
+     * Arrays save all their ints in one block on the heap while ArrayLists don't.
+     * They save only references to the boxed ints in one block.
+     * All the references point to the boxed ints which are usually spread all over the heap.
+     * This leads to many cache misses with a big error:
+     *
+     * Benchmark                               Mode  Cnt    Score   Error  Units
+     * JMHSample_39_MemoryAccess.sumArray      avgt   25   50.393 ± 0.407  ms/op
+     * JMHSample_39_MemoryAccess.sumArrayList  avgt   25  299.868 ± 1.591  ms/op
```

Comment:
It maybe can be even more educational if the benchmarks will expose how *unstable* such measurements are.

Having a third benchmark with the same `ArrayList` with the same `Integer` objects, but **shuffled** yields different result in the middle between these two. The educational hint here is that it can be GC who can do the actual shuffle

Reply:
My machine, `N = 20_000_000`:

```
Benchmark                          Mode  Cnt    Score    Error  Units
MemoryAccess.sumArray              avgt    5    7.296 ±  0.276  ms/op
MemoryAccess.sumArrayList          avgt    5   45.077 ±  1.787  ms/op
MemoryAccess.sumArrayListShuffled  avgt    5  544.628 ± 64.948  ms/op
```

Reply:
Yes, @mmirwaldt, do that ^ ;)

Reply:
The next level would be a multi-thread benchmark to show mutual impact of different threads and limitation of scalability due to hitting of the memory bandwidth limits for L1, L2, LLC, and RAM when the total throughput stops increasing after some small number of threads.

Reply:
@qwwdfsad:
 I can't reproduce your results with the shuffled array list with 20_000_000 integers.
 
 These are my results from an idle linux server:
 `
Benchmark                                                                Mode  Cnt   Score      Error    Units
JMHSample_39_MemoryAccess.sumArray                   avgt   25   4.880  ± 0.007   ms/op
JMHSample_39_MemoryAccess.sumArrayList              avgt   25  35.848 ± 0.170  ms/op
JMHSample_39_MemoryAccess.sumArrayListShuffled  avgt   25  35.838 ± 0.088  ms/op
`

You can find my code here on a separate branch which is not to be submitted as PR:
https://github.com/mmirwaldt/jmh/blob/samples/JMHSample_39_MemoryAccess_2/jmh-samples/src/main/java/org/openjdk/jmh/samples/JMHSample_39_MemoryAccess.java

What did you do different in your code?

Reply:
Yeah, that is a good idea.

I have added the shuffled array list and got similar results on an idle linux server:

```
Benchmark                                       Mode  Cnt    Score   Error  Units
JMHSample_39_MemoryAccess.sumArray              avgt   25    4.887 ± 0.008  ms/op
JMHSample_39_MemoryAccess.sumArrayList          avgt   25   35.765 ± 0.112  ms/op
JMHSample_39_MemoryAccess.sumArrayListShuffled  avgt   25  169.301 ± 1.064  ms/op
```